floyd/cli/run.py

```diff
@@ -66,8 +71,22 @@ def process_data_ids(data):
     return True, data_ids
 
 
-def validate_env(env, instance_type):
-    arch = INSTANCE_ARCH_MAP[instance_type]
+def resolve_final_instance_type(instance_type_override, yaml_str, task, cli_default):
+    if instance_type_override:
+        return instance_type_override
+    if yaml_str:
+        yaml_config = yaml.safe_load(yaml_str)
+        if task:
+            machine = yaml_config['task'][task].get('machine')
+            if machine:
+                return INSTANCE_TYPE_MAP[machine]
+        machine = yaml_config.get('machine')
+        if machine:
+            return INSTANCE_TYPE_MAP[machine]
```

Comment:
The `if machine` check seems to be duplicated, can they be refactored out? That could also improve readability.